prereise/gather/flexibilitydata/bus/bus_data.py

```diff
@@ -0,0 +1,339 @@
+
+        time.sleep(0.02)
+
+    with open(os.path.join(cache_path, "bus_fips.pkl"), "wb") as fh:
+        pkl.dump(bus_fips_dict, fh)
+
+
+def cleanup_zip(zipdict):
+    """Try to cleanup a zip dictionary obtained using online query by converting
+    to 5-digit integers. Several possible mis-format are considered
+
+    :param dict zipdict: a dictionary containing raw zip-code of buses
+    :return: (*dict*) -- a dictionary containing 5-digit zip codes
+    """
+    for i in range(len(zipdict["zip"])):
+        try:
+            zipdict["zip"][i] = int(zipdict["zip"][i])
+        except Exception:
+            if ":" in zipdict["zip"][i]:
+                zipdict["zip"][i] = int(zipdict["zip"][i].split(":")[0])
+
+            elif "-" in zipdict["zip"][i]:
+                zipdict["zip"][i] = int(zipdict["zip"][i].split("-")[0])
+
+            elif "‑" in zipdict["zip"][i]:
+                zipdict["zip"][i] = int(zipdict["zip"][i].split("‑")[0])
+
+            elif " " in zipdict["zip"][i]:
+                success = 0
+                for j in zipdict["zip"][i].split(" "):
+                    if len(j) == 5:
+                        zipdict["zip"][i] = int(j)
+                        success = 1
+                if success == 0:
+                    zipdict["zip"][i] = -1
+
+            else:
+                try:
+                    zipdict["zip"][i] = int(zipdict["zip"][i][0:5])
+                except Exception:
+                    zipdict["zip"][i] = -1
+
+    return zipdict
+
+
+def get_bus_zip(bus_pos, cache_path, start_idx=0):
+    """Try to get ZIP of each bus in a case mat using geopy
+    Can take hours to run, save to cache file for future use
+
+    :param numpy.ndarray bus_pos: a list of (busID, lat, lon) stored as a numpy array
+    :param str cache_path: folder to store processed cache files
+    :param int start_idx: pointer to the index of a bus to start query from
+    """
+    bus_num = bus_pos.shape[0]
+    bus_zip_dict = {
+        "busid": bus_pos[:, 0],
+        "latitude": bus_pos[:, 1],
+        "longitude": bus_pos[:, 2],
+        "zip": [0] * bus_num,
+    }
+
+    geocoder = geopy.Nominatim(user_agent="BES")
+    reverse = RateLimiter(
+        geocoder.reverse, min_delay_seconds=0.05, return_value_on_exception=None
+    )
+
+    def get_zip_code(lat, lon):
+        location = reverse("{}, {}".format(lat, lon))
+        if location is not None:
+            address = location.raw["address"]
+        else:
+            return -1
+
+        if "postcode" in address.keys():
+            return address["postcode"]
+        else:
+            return -1
+
+    for i in range(start_idx, bus_num):
+        bus_zip_dict["zip"][i] = int(
+            get_zip_code(bus_zip_dict["latitude"][i], bus_zip_dict["longitude"][i])
+        )
+
+    bus_zip_dict = cleanup_zip(bus_zip_dict)
+
+    with open(os.path.join(cache_path, "bus_zip.pkl"), "wb") as fh:
+        pkl.dump(bus_zip_dict, fh)
+
+
+def ll2uv(lon, lat):
```

Comment:
> Shall we put function `ll2uv`, `angular_distance`, `find_closest_neighbor` into a helper module instead?

It does not seem to be needed elsewhere. But say if I were to put them separately, can I just put it in the same folder as all the other code, just under a separate filename?

Reply:
Quick question. The functions are already defined in `powersimdata.utility.distance`, should we import them instead?

Reply:
> Quick question. The functions are already defined in `powersimdata.utility.distance`, should we import them instead?

right, I copied the functions from that script. but prereise is not dependent upon powersimdata right? can we just assume it has been installed?

Reply:
PreREISE depends on PowerSimData. You can see that in the Pipfile or requirements.txt file

Reply:
@rouille The original idea was making `PreREISE` independent of `PowerSimData`, which was a parallel effort for the time being. The strategy was avoiding to induce further dependencies and removing existing ones by refactor in the meantime (generating normalized profiles rather than in MWh was also part of that effort). Do we still wanna do it?

Reply:
I see. It is up to you then. If you redefine the function, then you should rewrite the tests. My only worry is that these functions might be used (or will be used) in PostREISE and we will end up with the same set of functions to calculate distance in all 3 packages

Reply:
@dongqi-wu Let's import these functions from `PowerSimData` for now then, we will take care of the decoupling refactor in the future.

Reply:
> @dongqi-wu Let's import these functions from `PowerSimData` for now then, we will take care of the decoupling refactor in the future.

sure, I will change to that for the coming commit.